code/artificial_intelligence/src/a_star/astar.cpp

```diff
@@ -0,0 +1,242 @@
+#include <list>
+#include <algorithm>
+#include <iostream>
+ 
+class point 
+{
+public:
+    point (int a = 0, int b = 0) 
+    { 
+        x = a; y = b; 
+    }
+
+    bool operator == (const point& o) 
+    { 
+        return o.x == x && o.y == y; 
+    }
+
+    point operator + (const point& o) 
+    { 
+        return point (o.x + x, o.y + y); 
+    }
+
+    int x, y;
+};
+ 
+class map 
+{
+public:
+    map() 
+    {
+        char t[8][8] = 
+        {
+            {0, 0, 0, 0, 0, 0, 0, 0}, {0, 0, 0, 0, 0, 0, 0, 0},
+            {0, 0, 0, 0, 1, 1, 1, 0}, {0, 0, 1, 0, 0, 0, 1, 0},
+            {0, 0, 1, 0, 0, 0, 1, 0}, {0, 0, 1, 1, 1, 1, 1, 0},
+            {0, 0, 0, 0, 0, 0, 0, 0}, {0, 0, 0, 0, 0, 0, 0, 0}
+        };
+
+        w = h = 8;
+        for (int r = 0; r < h; ++r)
+            for(int s = 0; s < w; ++s)
+                m[s][r] = t[r][s];
+    }
+
+    int operator() (int x, int y) 
+    { 
+        return m[x][y]; 
+    }
+
+    char m[8][8];
```

Comment:
I feel like m should be private, and w and h should be accessed through member functions. Also, don't use one letter variable names.